Added the ndcg metric [WIP]

kamalojasv181 opened the pull request:
Related #2631

Description: This is the implementation [WIP] for the NDCG metric.

Check list:

- [X] Create a basic implementation. 
- [x] Handle rating ties.
- [x] Write and parameterize tests.
- [ ] Add comments about its use
- [X] Add exponential implementation
- [ ] Add Docs

kamalojasv181 committed:
Added the ndcg metric [WIP]

kamalojasv181 committed:
Merge branch 'master' into ndcg

kamalojasv181 committed:
added GPU support, corrected mypy errors, and minor fixes

kamalojasv181 committed:
Merge branch 'ndcg' of https://github.com/kamalojasv181/ignite into ndcg

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+        self,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+        k: Union[int, None] = None,
```
```suggestion
        k: Optional[int] = None,
```

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+
+        self.num_examples += y_pred.shape[0]
+
+    def compute(self) -> torch.float32:  # type: ignore
```
```suggestion
    def compute(self) -> float:  # type: ignore
```

kamalojasv181 replied:
This function returns a torch tensor. Do you want me to return a float or just set the return type as a float so that mypy does not throw an error?

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+
+        y_pred, y = output[0], output[1]
```
```suggestion
        y_pred, y = output[0].detach(), output[1].detach()
```

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+        self.ngcd = torch.tensor(0, dtype=torch.float32, device=device)
+
+    def _dcg_sample_scores(
+        self, y_true: torch.Tensor, y_score: torch.Tensor, k: Union[int, None] = None, log_base: Union[int, float] = 2
```
```suggestion
        self, y_true: torch.Tensor, y_score: torch.Tensor, k: Optional[int] = None, log_base: Union[int, float] = 2
```

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+    def reset(self) -> None:
+
+        self.num_examples = 0
+        self.ngcd = torch.tensor(0, device=self._device)
```
```suggestion
        self.ngcd = torch.tensor(0.0, device=self._device)
```

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+
+        normalizing_gain = self._dcg_sample_scores(y_true, y_true, k, log_base=log_base)
+
+        all_irrelevant = normalizing_gain == 0
+
+        normalized_gain = torch.div(gain[~all_irrelevant], normalizing_gain[~all_irrelevant])
```
```suggestion
        all_relevant = normalizing_gain != 0

        normalized_gain = torch.div(gain[all_relevant], normalizing_gain[all_relevant])
```
Maybe, we can simplify that a bit

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+        k: Union[int, None] = None,
+        log_base: Union[int, float] = 2,
+    ):
+        super(NDCG, self).__init__(output_transform=output_transform, device=device)
```
In practice we call `super` in the last as it calls `reset()`

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+        self.log_base = log_base
+        self.k = k
+        self.num_examples = 0
+        self.ngcd = torch.tensor(0, dtype=torch.float32, device=device)
```
```suggestion
        self.ngcd = torch.tensor(0.0, device=device)
```

vfdev-5 reviewed (commented):
@kamalojasv181 thanks for the PR. I left few comments on few points, but I haven't yet explored the code to compute ndcg.

Can you please write few tests vs scikit-learn as ref ?

kamalojasv181 committed:
Incorporated the suggested changes

kamalojasv181 committed:
Fixed mypy error

vfdev-5 commented on ignite/metrics/ndcg.py:
```diff
@@ -0,0 +1,81 @@
+
+        super(NDCG, self).__init__(output_transform=output_transform, device=device)
+
+    def _dcg_sample_scores(
```
Let's put dcg and `ndcg_sample_scores` functions out of NDCG class just define them before the class, such that they could be also used alone.

vfdev-5 reviewed (commented):
Thanks a lot for the update @kamalojasv181 !
I left few other comments on the implementation and the API.

Let's also start working on docs and tests

kamalojasv181 commented:
Thanks for all the feedback. I will revert with a pull request addressing everything!

vfdev-5 commented:
> Thanks for all the feedback. I will revert with a pull request addressing everything!

You can just continue working with this pull request, no need to revert anything.

kamalojasv181 committed:
Fixed bugs in NDCG and added tests for output and reset

kamalojasv181 committed:
Fixed mypy error

kamalojasv181 committed:
Added the exponential form on https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Discounted_Cumulative_Gain

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+
+        y, y_pred = output[0].detach(), output[1].detach()
```
```suggestion
        y_pred, y  = output[0].detach(), output[1].detach()
```

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+
+        y, y_pred = output[0].detach(), output[1].detach()
+        gain = self._ndcg_sample_scores(y, y_pred, k=self.k, log_base=self.log_base)
+        self.ndcg = torch.add(self.ndcg, torch.sum(gain))
```
Can we use plain ops like `+` instead of torch.add ?
```suggestion
        self.ndcg += torch.sum(gain)
```

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+        return discounted_gains
+
+    def _ndcg_sample_scores(
+        self, y_true: torch.Tensor, y_score: torch.Tensor, k: Optional[int] = None, log_base: Union[int, float] = 2
```
Let's use pytorch/ignite convention on args: (y_pred, y_true) instead of (y_true, y_pred)

vfdev-5 replied:
yes, please

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+        self.exponential = exponential
+        super(NDCG, self).__init__(output_transform=output_transform, device=device)
+
+    def _dcg_sample_scores(
```
Let's put _dcg_sample_scores and _ndcg_sample_scores outside of the class as plain methods in ignite/metrics/recsys/ndcg.py

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+        self, y_true: torch.Tensor, y_score: torch.Tensor, k: Optional[int] = None, log_base: Union[int, float] = 2
+    ) -> torch.Tensor:
+
+        discount = (
+            torch.div(torch.log(torch.arange(y_true.shape[1]) + 2), torch.log(torch.tensor(log_base)))
+            .pow(-1)
+            .to(self._device)
+        )
```
```suggestion
        discount = torch.log(torch.tensor(log_base)) / torch.log(torch.arange(y_true.shape[1]) + 2)
        discount = discount.to(self._device)
```

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+        gain = self._dcg_sample_scores(y_true, y_score, k, log_base=log_base)
+        normalizing_gain = self._dcg_sample_scores(y_true, y_true, k, log_base=log_base)
+        all_relevant = normalizing_gain != 0
+        normalized_gain = torch.div(gain[all_relevant], normalizing_gain[all_relevant])
```
```suggestion
        normalized_gain = gain[all_relevant] / normalizing_gain[all_relevant]
```

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,73 @@
+        exponential: bool = False,
+    ):
+
+        self.log_base = log_base
```
Let's assert that provided log_base is not negative and not 1

vfdev-5 replied:
@kamalojasv181 we usually do:
```python
if log_base <=0 or log_base == 1:
    raise ValueError("...")
```
instead of `assert ...`

kamalojasv181 committed:
Corrected true, pred order and corresponding tests

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,67 @@
+        ),
+    ],
+)
+@pytest.mark.parametrize("k", [None, 2, 3])
```
Let's try to think how can we test for other options like : 
- exponential
- log_base

For exponential, maybe we still can use sklearn and rescale inputs ?

kamalojasv181 replied:
Thats right

kamalojasv181 replied:
For log_base I can rewrite the sklearn function?

vfdev-5 replied:
Yes, we can try that

kamalojasv181 replied:
Sure, will have a look

vfdev-5 reviewed (commented):
Thanks for the updates @kamalojasv181 !
I have few other code update suggestions.

kamalojasv181 committed:
Added ties case, exponential tests, log_base tests, corresponding tests and other suggested changes

kamalojasv181 committed:
Added GPU check on top

kamalojasv181 committed:
Put tensors on GPU inside the function to pervent error

kamalojasv181 commented:
@vfdev-5 there are a bunch of things I did in this commit:
1) Put the data on GPU for tests
2) Replaced torch.add with +.
3) Changed the naming and ordering as per ignite as (y_pred, y_true ....)
4) Added function to handle ties and wrote corresponding tests(A part of it is un-vectorized, TODO: Think of implementation to vectorise it.)
5) Added check for log_base > 0 and not equal to 1.
6) Added tests for log_base and exponential. 
7) Separated the _ndcg_smaple_score and _dcg_smaple_score from the NDCG class. 
NOTE: log_base tests could have been done with the other tests, but I put them separately so that if sklearn changes their code, then we know about it (as the tests would be failing)

If there is anything else, lemme know before I can finally add some comments against the class and documentation.

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,115 @@
+        if self.exponential:
+            y_true = 2 ** y_true - 1
+
+        gain = _ndcg_sample_scores(y_pred, y_true, k=self.k, log_base=self.log_base, device=self._device)
```
Let's remove device arg from `_ndcg_sample_scores` and use `y_pred.device` internally. Once `gain` is returned we put it to `self._device` before adding to `self.ndcg`

vfdev-5 replied:
You do not pass `ignore_ties` arg to `_ndcg_sample_scores` ?

kamalojasv181 replied:
I do

kamalojasv181 replied:
I think you checked some previous commit

kamalojasv181 replied:
I am sorry, I forgot to push that

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,187 @@
+    ],
+)
+@pytest.mark.parametrize("k", [None, 2, 3])
+def test_output_gpu_ignore_ties(y_pred, y_true, k):
```
Maybe `ignore_ties` can be a parameter like `k`

kamalojasv181 replied:
Actually we dont want to test the examples with ties with `ignore_ties = True`. Thats why

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,187 @@
+    ndcg = NDCG(k=k, device=device, ignore_ties=True)
+    ndcg.update([y_pred, y_true])
+    result_ignite = ndcg.compute()
+    result_sklearn = ndcg_score(y_true.cpu().numpy(), y_pred.cpu().numpy(), k=k)
```
You have to pass `ignore_ties` arg to `ndcg_score` as well

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -0,0 +1,115 @@
+
+    else:
+        discount_cumsum = torch.cumsum(discount, dim=-1)
+        discounted_gains = torch.tensor(
+            [_tie_averaged_dcg(y_p, y_t, discount_cumsum, device) for y_p, y_t in zip(y_pred, y_true)], device=device
+        )
```
So, there is no way to make it vectorized == without for-loop ?

kamalojasv181 replied:
I havent checked yet. For now I have added this implementation. It's a TODO

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,187 @@
+
+@pytest.mark.parametrize(
+    "y_pred, y_true",
+    [
+        (torch.tensor([[0.1, 0.2, 0.3, 0.4, 0.5]]), torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0]])),
+        (torch.tensor([[3.7, 4.8, 3.9, 4.3, 4.9]]), torch.tensor([[2.9, 5.6, 3.8, 7.9, 6.2]])),
+        (
+            torch.tensor([[0.1, 0.2, 0.3, 0.4, 0.5], [3.7, 4.8, 3.9, 4.3, 4.9], [3.7, 4.8, 3.9, 4.3, 4.9]]),
+            torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0], [1.2, 4.5, 8.9, 5.6, 7.2], [2.9, 5.6, 3.8, 7.9, 6.2]]),
+        ),
+        (torch.tensor([[3.7, 3.7, 3.7, 3.7, 3.7]]), torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0]])),
+        (
+            torch.tensor([[3.7, 3.7, 3.7, 3.7, 3.7], [3.7, 3.7, 3.7, 3.7, 3.9]]),
+            torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0], [1.0, 2.0, 3.0, 4.0, 5.0]]),
+        ),
+    ],
```
If we generate this data multiple times, maybe we could refactor it into a function the return a list of these test cases ?

kamalojasv181 replied:
Sure

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,187 @@
+    ],
+)
+@pytest.mark.parametrize("k", [None, 2, 3])
+def test_exponential(y_pred, y_true, k):
```
I'd also prefer to have `exponential` as pytest parametrize such that we have a single code to check everything. As for sklearn data we can do:
```python
if exponential:
    y_true = 2 ** y_true - 1

result_sklearn = ndcg_score(y_true.numpy(), y_pred.numpy(), k=k)
```

kamalojasv181 replied:
Alright!

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,187 @@
+
+@pytest.mark.parametrize("log_base", [2, 3, 10])
+def test_log_base(log_base):
+    def _ndcg_sample_scores(y_true, y_score, k=None, ignore_ties=False):
+
+        gain = _dcg_sample_scores(y_true, y_score, k, ignore_ties=ignore_ties)
+        normalizing_gain = _dcg_sample_scores(y_true, y_true, k, ignore_ties=True)
+        all_irrelevant = normalizing_gain == 0
+        gain[all_irrelevant] = 0
+        gain[~all_irrelevant] /= normalizing_gain[~all_irrelevant]
+        return gain
+
+    def ndcg_score_with_log_base(y_true, y_score, *, k=None, sample_weight=None, ignore_ties=False, log_base=2):
+
+        gain = _ndcg_sample_scores(y_true, y_score, k=k, ignore_ties=ignore_ties)
+        return np.average(gain, weights=sample_weight)
```
Let's put this outside of `test_log_base`

kamalojasv181 committed:
Improved tests and minor bugfixes

kamalojasv181 committed:
Removed device hyperparam from _ndcg_smaple_scores

kamalojasv181 committed:
Skipped GPU tests for CPU only systems

vfdev-5 commented on ignite/metrics/recsys/ndcg.py:
```diff
@@ -85,12 +85,13 @@ def __init__(
 
-        assert log_base != 1 or log_base <= 0, f"Illegal value {log_base} for log_base"
+        if log_base == 1 or log_base <= 0:
+            raise ValueError(f"Illegal value {log_base} for log_base")
```
```suggestion
            raise ValueError(f"Argument log_base should positive and not equal one,but got {log_base}")
```

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -8,62 +8,78 @@
     result_ignite = ndcg.compute()
-    result_sklearn = ndcg_score(y_true.numpy(), y_pred.numpy(), k=k)
+
+    if try_ignore_ties:
+        ndcg = NDCG(k=k, device=device, exponential=exponential, ignore_ties=True)
+        ndcg.update([y_pred, y_true])
+        result_ignite_ignore_ties = ndcg.compute()
```
Why can't we code it as 
```python
ndcg = NDCG(k=k, device=device, exponential=exponential, ignore_ties=try_ignore_ties)
ndcg.update([y_pred, y_true])
result_ignite = ndcg.compute()

result_sklearn = ndcg_score(y_true.numpy(), y_pred.numpy(), k=k, ignore_ties=try_ignore_ties)
```
?

kamalojasv181 replied:
We can. But we will end up computing the `ignore_ties = False` twice in case of `try_ignore_ties = True` which is inefficient and in my personal opinion a little confusing. However if you say so, I can do it this way

kamalojasv181 committed:
Changed Error message

vfdev-5 commented:
Hi @kamalojasv181 can you continue on this PR ?

kamalojasv181 committed:
Merge branch 'master' of https://github.com/pytorch/ignite into ndcg

kamalojasv181 committed:
Merge branch 'pytorch:master' into ndcg

kamalojasv181 committed:
Made tests randomised from deterministic and introduced ignore_ties_flag to reduce duplicate code

kamalojasv181 committed:
Merge branch 'ndcg' of https://github.com/kamalojasv181/ignite into ndcg

kamalojasv181 commented:
@vfdev-5 can you please review the changes ?

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,145 @@
+
+@pytest.mark.parametrize("k", [None, 2, 3])
+@pytest.mark.parametrize("exponential", [True, False])
+@pytest.mark.parametrize("ignore_ties_flag", [True, False])
+@pytest.mark.parametrize("ignore_ties_input", [True, False])
```
Maybe, we can rename them for better understanding like :
```
ignore_ties_flag -> ignore_ties
ignore_ties_input -> replacement
```

On the other hand, if we have ties in the input data but we compute metrics with `ignore_ties=True`, this would lead to incorrect computation ?
Maybe, we can parametrize without this test case:
```suggestion
@pytest.mark.parametrize("ignore_ties, replacement", [(True, False), (False, True), (False, False)])
```

kamalojasv181 replied:
In that case I just return out of the function without performing any tests. But sure, this is better.

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,145 @@
+@pytest.mark.parametrize("ignore_ties_flag", [True, False])
+@pytest.mark.parametrize("ignore_ties_input", [True, False])
+@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="Skip if no GPU")
+def test_output_gpu(test_case, k, exponential, ignore_ties_flag, ignore_ties_input):
```
Let's rename this test : test_output_gpu -> test_output_on_cuda such that it would run on GPU: https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#codebase-structure

vfdev-5 commented:
@kamalojasv181 thanks for the updates, I left few nits.
Next step is to write few tests for distributed configuration. @puhuk please provide few guidelines here on how to write them. Thanks !

kamalojasv181 committed:
Merge branch 'master' of https://github.com/pytorch/ignite into ndcg

kamalojasv181 committed:
Merge branch 'pytorch:master' into ndcg

kamalojasv181 committed:
Changed test name to test_output_cuda from test_output_gpu

kamalojasv181 committed:
Merge branch 'ndcg' of https://github.com/kamalojasv181/ignite into ndcg

kamalojasv181 committed:
Merge branch 'master' into ndcg

kamalojasv181 committed:
Changed variable names to replacement and ignore_ties and removed redundat tests

kamalojasv181 committed:
Changed variable names to replacement and ignore_ties and removed redundat tests

kamalojasv181 committed:
Merge branch 'master' of https://github.com/pytorch/ignite into ndcg

kamalojasv181 committed:
Merge branch 'ndcg' of https://github.com/kamalojasv181/ignite into ndcg

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,143 @@
+from ignite.metrics.recsys.ndcg import NDCG
+
+
+@pytest.fixture(params=[item for item in range(6)])
+def test_case(request):
+
+    return [
+        (torch.tensor([[3.7, 4.8, 3.9, 4.3, 4.9]]), torch.tensor([[2.9, 5.6, 3.8, 7.9, 6.2]])),
+        (
+            torch.tensor([[3.7, 3.7, 3.7, 3.7, 3.7], [3.7, 3.7, 3.7, 3.7, 3.9]]),
+            torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0], [1.0, 2.0, 3.0, 4.0, 5.0]]),
+        ),
+    ][request.param % 2]
```
@kamalojasv181 seems like test cases 4, 5, 6 are redundant, let's reduce the number of test cases:
```suggestion
@pytest.fixture(params=[item for item in range(3)])
def test_case(request):

    return [
        (torch.tensor([[3.7, 4.8, 3.9, 4.3, 4.9]]), torch.tensor([[2.9, 5.6, 3.8, 7.9, 6.2]])),
        (
            torch.tensor([[3.7, 3.7, 3.7, 3.7, 3.7], [3.7, 3.7, 3.7, 3.7, 3.9]]),
            torch.tensor([[1.0, 2.0, 3.0, 4.0, 5.0], [1.0, 2.0, 3.0, 4.0, 5.0]]),
        ),
    ][request.param]
```

kamalojasv181 replied:
I added them so that we can test on multiple random inputs. But if you say, I can remove them.

vfdev-5 replied:
Oh, I see what you mean. That's true that we generate preds with multinomial. OK, let's keep your code.

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,143 @@
+
+    y_pred = torch.multinomial(y_pred_distribution, 5, replacement=replacement)
+
+    if not replacement and ignore_ties:
+        return
```
As we are doing `@pytest.mark.parametrize("ignore_ties, replacement", [(True, False), (False, True), (False, False)])` we can remove the code like this one:
```python
    if not replacement and ignore_ties:
        return
```

puhuk commented:
@vfdev-5 @kamalojasv181 

Belows are checklist for test in ddp configuration



- [ ] Data generating with different random seed per each process
    - Like `torch.random_seed(12 + rank + i)`
- [ ] Update data with batch style
    - `y_pred[i * batch_size : (i+1) * batch_size]`
- [ ] Check generated data from each processes are gathered (after running Engine)
   -  ```python 
         engine.run(data=data, max_epochs=n_epochs)
         y_true = idist.all_gather(y_true)
         y_preds = idist.all_gather(y_preds)
      ```
- [ ] Check calculated metric is same as reference
   - ` pytest.approx(calculated_value) == reference_value`

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,143 @@
+
+    y_pred = torch.multinomial(y_pred_distribution, 5, replacement=replacement)
+
+    if not replacement and ignore_ties:
+        return
```
Same here

kamalojasv181 committed:
Removed redundant test cases and removed the redundant if statement

kamalojasv181 committed:
Added distributed tests, added multiple test cases corresponding to one multiomial distribution

kamalojasv181 committed:
Made the tests wsork on in ddp configuration

kamalojasv181 committed:
Merge branch 'master' of https://github.com/pytorch/ignite into ndcg

kamalojasv181 committed:
Merge branch 'pytorch:master' into ndcg

kamalojasv181 committed:
Merge branch 'ndcg' of https://github.com/kamalojasv181/ignite into ndcg

vfdev-5 commented on tests/ignite/metrics/test_ndcg.py:
```diff
@@ -0,0 +1,266 @@
+        y_preds = torch.rand((n_iters * batch_size, n_items)).to(device)
+
+        def update(_, i):
+            return (
+                [v for v in y_preds[i * batch_size : (i + 1) * batch_size, ...]],
+                [v for v in y_true[i * batch_size : (i + 1) * batch_size]],
+            )
```
Why do you return tuple of 2 lists instead of a tuple of two tensors ?

kamalojasv181 replied:
I have taken inspiration from the code provided by @puhuk . Here each element of the list is a batch. We feed our engine one batch at a time; hence using a list is also ok. To maintain uniformity across the code, I have kept it this way.

vfdev-5 replied:
Oh, I see, he provided a wrong link. Yes, in accuracy we also a test on list of tensors and numbers but this is untypical. Here is a typical example https://github.com/pytorch/ignite/blob/26f7cec873b382a500204aa0121bff1661248462/tests/ignite/metrics/test_accuracy.py#L412-L463

vfdev-5 commented:
@sadra-barikbin can you check why tests/ignite/metrics/test_precision.py::test_binary_input[None] is failing systematically ?

```
E       assert array([1.]) == approx([1.0 ± 1.0e-06, 0.0 ± 1.0e-12])
E         Impossible to compare arrays with different shapes.
E         Shapes: (2,) and (1,)

tests/ignite/metrics/test_precision.py:130: AssertionError
```

kamalojasv181 committed:
Merge branch 'pytorch:master' into ndcg

kamalojasv181 committed:
Returning tuple of two tensors instead of tuple of list of tensors

puhuk commented:
@kamalojasv181 Hi, do you need any help to finalize this PR? Please feel free to let me and @vfdev-5  know :)

ili0820 commented:
Any updates on this ? If it is not finished yet, I'd love to contribute @vfdev-5

vfdev-5 commented:
Yes, this PR is not finished, unfortunately. @ili0820 if you can help with getting it landed it would be great!

ili0820 commented:
after reviewing all the comments,
As far as I understand,  the testing part is the last and only problem that has been not resolved yet,
[this](https://github.com/pytorch/ignite/pull/2632#discussion_r928635489) seems resolved and 
[this](https://github.com/pytorch/ignite/pull/2632#discussion_r959502399) is the problem.
Is this correct? @vfdev-5

vfdev-5 commented:
> after reviewing all the comments, As far as I understand, the testing part is the last and only problem that has been not resolved yet, [this](https://github.com/pytorch/ignite/pull/2632#discussion_r928635489) seems resolved and [this](https://github.com/pytorch/ignite/pull/2632#discussion_r959502399) is the problem. Is this correct? @vfdev-5

@ili0820 yes, it remains few things here:
- https://github.com/pytorch/ignite/pull/2632/files#r930048810
- test for distributed configuration

In case if you are not familiar with DDP, please check: https://pytorch-ignite.ai/tutorials/advanced/01-collective-communication/. As for testing best practices, we would like to use now `distributed` fixture as here: https://github.com/pytorch/ignite/blob/f2b11832549b5b6623ea6791e71781f930336c93/tests/ignite/metrics/test_ssim.py#L226
if you have any questions, you can reach out to us on [Discord](https://pytorch-ignite.ai/chat) in #start-contributing channel.

vfdev-5 commented:
Closing in favor of https://github.com/pytorch/ignite/pull/3608
